lib/tree.rb

```diff
@@ -1,66 +1,135 @@
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
-
-   def initialize(key, val)
+  
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(log (n))
+  # Space Complexity: O(log (n))
   def add(key, value)
-    raise NotImplementedError
+    #recursive
+    @root = add_helper(@root, key, value)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key, value) if current_node.nil?
+    
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    
+    return current_node
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def find(key)
-    raise NotImplementedError
+    helper_find(@root, key)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def helper_find(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+    
+    if key < current_node.key
+      helper_find(current_node.left, key)
+    else
+      helper_find(current_node.right, key)
+    end
+  end
+  
+  # Time Complexity: o(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(current_node, list)
+    if current_node.nil?
+      return list
+    end
+    
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value}
+    inorder_helper(current_node.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(current_node, list)
+    if current_node.nil?
+      return list
+    end
+    
+    list << { key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def postorder_helper(current_node, list)
+    if current_node.nil?
+      return list
+    end
+    
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << { key: current_node.key, value: current_node.value}
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
You're visiting each node so the time complexity is right, but the space complexity (for a balanced tree) is O(log n ) because your call stack will never get greater than the height of the tree.